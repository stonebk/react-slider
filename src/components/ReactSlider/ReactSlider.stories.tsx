import { useState } from 'react';
import { ReactSlider } from './ReactSlider';

const meta = {
  title: 'ReactSlider',
  component: ReactSlider,
  tags: ['autodocs'],
};

export default meta;

const defaults = {
  min: 0,
  max: 100,
  step: 1,
  pageFn: (step) => step * 10,
  minDistance: 0,
  defaultValue: 0,
  orientation: 'horizontal',
  className: 'react-slider',
  thumbClassName: 'react-slider__thumb',
  thumbActiveClassName: 'react-slider__thumb--active',
  trackClassName: 'react-slider__track',
  markClassName: 'react-slider__mark',
  withTracks: true,
  pearling: false,
  disabled: false,
  snapDragDisabled: false,
  invert: false,
  marks: [],
  renderThumb: ({ key, ...props }) => <div key={key} {...props} />,
  renderTrack: ({ key, ...props }) => <div key={key} {...props} />,
  renderMark: ({ key, ...props }) => <span key={key} {...props} />,
};

/**
 * Single slider, similar to `<input type="range" defaultValue={0} />`
 */

export const DefaultSlider = {
  args: {
    ...defaults,
    className: 'react-slider--horizontal',
    renderThumb: (props, state) => <div {...props}>{state.valueNow}</div>,
  },
};

export const SliderWithMarks = {
  args: {
    ...defaults,
    className: 'react-slider--horizontal',
    marks: true,
    min: 0,
    max: 9,
    renderThumb: (props, state) => <div {...props}>{state.valueNow}</div>,
  },
};

export const InvertedSliderWithMarks = {
  args: {
    ...defaults,
    className: 'react-slider--horizontal',
    marks: [5, 6, 7, 8, 9],
    min: 0,
    max: 9,
    invert: true,
    renderThumb: (props, state) => <div {...props}>{state.valueNow}</div>,
  },
};

export const DoubleSlider = {
  args: {
    ...defaults,
    className: 'react-slider--horizontal',
    defaultValue: [0, 100],
    ariaLabel: ['Lower thumb', 'Upper thumb'],
    ariaValuetext: (state) => `Thumb value ${state.valueNow}`,
    pearling: true,
    minDistance: 10,
    renderThumb: (props, state) => <div {...props}>{state.valueNow}</div>,
  },
};

export const MultiSlider = {
  args: {
    ...defaults,
    className: 'react-slider--horizontal',
    defaultValue: [0, 50, 100],
    ariaLabel: ['Leftmost thumb', 'Middle thumb', 'Rightmost thumb'],
    pearling: true,
    minDistance: 10,
    renderThumb: (props, state) => <div {...props}>{state.valueNow}</div>,
  },
};

export const VerticalSlider = {
  args: {
    ...defaults,
    className: 'react-slider--vertical',
    defaultValue: [0, 50, 100],
    ariaLabel: ['Lowest thumb', 'Middle thumb', 'Top thumb'],
    orientation: 'vertical',
    invert: true,
    pearling: true,
    minDistance: 10,
    renderThumb: (props, state) => <div {...props}>{state.valueNow}</div>,
  },
};

export const VerticalSliderWithMarks = {
  args: {
    ...defaults,
    className: 'react-slider--vertical',
    defaultValue: [0, 50, 100],
    marks: 25,
    ariaLabel: ['Lowest thumb', 'Middle thumb', 'Top thumb'],
    orientation: 'vertical',
    invert: true,
    pearling: true,
    minDistance: 10,
    renderThumb: (props, state) => <div {...props}>{state.valueNow}</div>,
  },
};

/**
 * Track changes with `onBeforeChange`, `onChange`, and `onAfterChange` event handlers.
 */

export const SliderWithEventHandlers = {
  args: {
    ...defaults,
    className: 'react-slider--horizontal',
    onBeforeChange: (value, index) =>
      console.log(`onBeforeChange: ${JSON.stringify({ value, index })}`),
    onChange: (value, index) =>
      console.log(`onChange: ${JSON.stringify({ value, index })}`),
    onAfterChange: (value, index) =>
      console.log(`onAfterChange: ${JSON.stringify({ value, index })}`),
    renderThumb: (props, state) => <div {...props}>{state.valueNow}</div>,
  },
};

/**
 * Using the `onChange` event handler, you can use the slider as a controlled component.
 */

export const SliderWithControlledValue = {
  args: {
    ...defaults,
    onBeforeChange: (value, index) =>
      console.log(`onBeforeChange: ${JSON.stringify({ value, index })}`),
    onChange: (value, index) =>
      console.log(`onChange: ${JSON.stringify({ value, index })}`),
    onAfterChange: (value, index) =>
      console.log(`onAfterChange: ${JSON.stringify({ value, index })}`),
    className: 'react-slider--horizontal',
    renderThumb: (props, state) => <div {...props}>{state.valueNow}</div>,
  },
  render: (args) => {
    // eslint-disable-next-line react-hooks/rules-of-hooks
    const [value] = useState([25, 50]);
    return <ReactSlider {...args} value={value} />;
  },
};
