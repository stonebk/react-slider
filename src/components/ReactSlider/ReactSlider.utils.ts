/**
 * To prevent text selection while dragging.
 * http://stackoverflow.com/questions/5429827/how-can-i-prevent-text-element-selection-with-cursor-drag
 */
export const pauseEvent = (e) => {
  if (e && e.stopPropagation) {
    e.stopPropagation();
  }
  if (e && e.preventDefault) {
    e.preventDefault();
  }
  return false;
};

export const stopPropagation = (e) => {
  if (e.stopPropagation) {
    e.stopPropagation();
  }
};

export const sanitizeInValue = (x) => {
  if (x == null) {
    return [];
  }
  return Array.isArray(x) ? x.slice() : [x];
};

export const prepareOutValue = (x) => {
  return x !== null && x.length === 1 ? x[0] : x.slice();
};

export const trimSucceeding = (length, nextValue, minDistance, max) => {
  for (let i = 0; i < length; i += 1) {
    const padding = max - i * minDistance;
    if (nextValue[length - 1 - i] > padding) {
      nextValue[length - 1 - i] = padding;
    }
  }
};

export const trimPreceding = (length, nextValue, minDistance, min) => {
  for (let i = 0; i < length; i += 1) {
    const padding = min + i * minDistance;
    if (nextValue[i] < padding) {
      nextValue[i] = padding;
    }
  }
};

export const addHandlers = (eventMap) => {
  Object.keys(eventMap).forEach((key) => {
    if (typeof document !== 'undefined') {
      document.addEventListener(key, eventMap[key], false);
    }
  });
};

export const removeHandlers = (eventMap) => {
  Object.keys(eventMap).forEach((key) => {
    if (typeof document !== 'undefined') {
      document.removeEventListener(key, eventMap[key], false);
    }
  });
};

export const trimAlignValue = (val, props) => {
  return alignValue(trimValue(val, props), props);
};

export const alignValue = (val, props) => {
  const valModStep = (val - props.min) % props.step;
  let alignedValue = val - valModStep;

  if (Math.abs(valModStep) * 2 >= props.step) {
    alignedValue += valModStep > 0 ? props.step : -props.step;
  }

  return parseFloat(alignedValue.toFixed(5));
};

export const trimValue = (val, props) => {
  let trimmed = val;
  if (trimmed <= props.min) {
    trimmed = props.min;
  }
  if (trimmed >= props.max) {
    trimmed = props.max;
  }

  return trimmed;
};
