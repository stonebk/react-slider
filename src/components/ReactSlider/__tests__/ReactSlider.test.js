import React from 'react';
import renderer from 'react-test-renderer';
import ReactSlider from '../ReactSlider';

window.ResizeObserver =
    window.ResizeObserver ||
    jest.fn().mockImplementation(() => ({
        disconnect: jest.fn(),
        observe: jest.fn(),
        unobserve: jest.fn(),
    }));

describe('<ReactSlider>', () => {
    it('can render', () => {
        const tree = renderer.create(<ReactSlider />).toJSON();
        expect(tree).toMatchSnapshot();
    });

    describe('event handlers', () => {
        beforeEach(() => {
            // https://github.com/facebook/jest/issues/890#issuecomment-209698782
            Object.defineProperty(document, 'addEventListener', {
                writable: true,
                value: jest.fn(),
            });
            Object.defineProperty(document, 'removeEventListener', {
                writable: true,
                value: jest.fn(),
            });
        });

        it('does not call any event handlers if the value does not change', () => {
            const onBeforeChange = jest.fn();
            const onChange = jest.fn();
            const onAfterChange = jest.fn();
            const testRenderer = renderer.create(
                <ReactSlider
                    onBeforeChange={onBeforeChange}
                    onChange={onChange}
                    onAfterChange={onAfterChange}
                    thumbClassName="test-thumb"
                    min={0}
                    step={1}
                />
            );

            const testInstance = testRenderer.root;
            const thumb = testInstance.findByProps({ className: 'test-thumb test-thumb-0 ' });

            const { addEventListener } = document;
            expect(addEventListener).not.toHaveBeenCalled();

            // simulate focus on thumb
            thumb.props.onFocus();

            expect(addEventListener).toHaveBeenCalledTimes(3);
            expect(addEventListener.mock.calls[0][0]).toBe('keydown');
            expect(addEventListener.mock.calls[1][0]).toBe('keyup');
            expect(addEventListener.mock.calls[2][0]).toBe('focusout');

            const onKeyDown = addEventListener.mock.calls[0][1];

            expect(onBeforeChange).not.toHaveBeenCalled();
            expect(onChange).not.toHaveBeenCalled();
            expect(onAfterChange).not.toHaveBeenCalled();

            // simulate keydown
            onKeyDown({ key: 'ArrowLeft', preventDefault: () => {} });

            expect(onBeforeChange).not.toHaveBeenCalled();
            expect(onChange).not.toHaveBeenCalled();
            expect(onAfterChange).not.toHaveBeenCalled();

            // simulate keydown
            onKeyDown({ key: 'Home', preventDefault: () => {} });

            expect(onBeforeChange).not.toHaveBeenCalled();
            expect(onChange).not.toHaveBeenCalled();
            expect(onAfterChange).not.toHaveBeenCalled();

            // simulate keydown
            onKeyDown({ key: 'PageDown', preventDefault: () => {} });

            expect(onBeforeChange).not.toHaveBeenCalled();
            expect(onChange).not.toHaveBeenCalled();
            expect(onAfterChange).not.toHaveBeenCalled();
        });

        it('calls onBeforeChange only once before onChange', () => {
            const onBeforeChange = jest.fn();
            const onChange = jest.fn();
            const testRenderer = renderer.create(
                <ReactSlider
                    onBeforeChange={onBeforeChange}
                    onChange={onChange}
                    thumbClassName="test-thumb"
                    min={0}
                    step={1}
                />
            );

            const testInstance = testRenderer.root;
            const thumb = testInstance.findByProps({ className: 'test-thumb test-thumb-0 ' });

            const { addEventListener } = document;
            expect(addEventListener).not.toHaveBeenCalled();

            // simulate focus on thumb
            thumb.props.onFocus();

            expect(addEventListener).toHaveBeenCalledTimes(3);
            expect(addEventListener.mock.calls[0][0]).toBe('keydown');
            expect(addEventListener.mock.calls[1][0]).toBe('keyup');
            expect(addEventListener.mock.calls[2][0]).toBe('focusout');

            const onKeyDown = addEventListener.mock.calls[0][1];

            expect(onBeforeChange).not.toHaveBeenCalled();
            expect(onChange).not.toHaveBeenCalled();

            // simulate keydown
            onKeyDown({ key: 'ArrowRight', preventDefault: () => {} });

            expect(onBeforeChange).toHaveBeenCalledTimes(1);
            expect(onBeforeChange).toHaveBeenCalledWith(0, 0);
            expect(onBeforeChange.mock.invocationCallOrder[0]).toBeLessThan(
                onChange.mock.invocationCallOrder[0]
            );
            expect(onChange).toHaveBeenCalledTimes(1);
            expect(onChange).toHaveBeenCalledWith(1, 0);

            // simulate keydown
            onKeyDown({ key: 'ArrowRight', preventDefault: () => {} });

            expect(onBeforeChange).toHaveBeenCalledTimes(1);
            expect(onBeforeChange).toHaveBeenCalledWith(0, 0);
            expect(onChange).toHaveBeenCalledTimes(2);
            expect(onChange).toHaveBeenCalledWith(2, 0);
        });

        it('calls onChange for every change', () => {
            const onChange = jest.fn();
            const testRenderer = renderer.create(
                <ReactSlider onChange={onChange} thumbClassName="test-thumb" min={0} step={1} />
            );

            const testInstance = testRenderer.root;
            const thumb = testInstance.findByProps({ className: 'test-thumb test-thumb-0 ' });

            const { addEventListener } = document;
            expect(addEventListener).not.toHaveBeenCalled();

            // simulate focus on thumb
            thumb.props.onFocus();

            expect(addEventListener).toHaveBeenCalledTimes(3);
            expect(addEventListener.mock.calls[0][0]).toBe('keydown');
            expect(addEventListener.mock.calls[1][0]).toBe('keyup');
            expect(addEventListener.mock.calls[2][0]).toBe('focusout');

            const onKeyDown = addEventListener.mock.calls[0][1];

            expect(onChange).not.toHaveBeenCalled();

            // simulate keydown
            onKeyDown({ key: 'ArrowRight', preventDefault: () => {} });

            expect(onChange).toHaveBeenCalledTimes(1);
            expect(onChange).toHaveBeenCalledWith(1, 0);

            // simulate keydown
            onKeyDown({ key: 'ArrowLeft', preventDefault: () => {} });

            expect(onChange).toHaveBeenCalledTimes(2);
            expect(onChange).toHaveBeenCalledWith(0, 0);
        });

        it('calls onAfterChange only once after onChange', () => {
            const onChange = jest.fn();
            const onAfterChange = jest.fn();
            const testRenderer = renderer.create(
                <ReactSlider
                    onChange={onChange}
                    onAfterChange={onAfterChange}
                    thumbClassName="test-thumb"
                    min={0}
                    step={1}
                />
            );

            const testInstance = testRenderer.root;
            const thumb = testInstance.findByProps({ className: 'test-thumb test-thumb-0 ' });

            const { addEventListener } = document;
            expect(addEventListener).not.toHaveBeenCalled();

            // simulate focus on thumb
            thumb.props.onFocus();

            expect(addEventListener).toHaveBeenCalledTimes(3);
            expect(addEventListener.mock.calls[0][0]).toBe('keydown');
            expect(addEventListener.mock.calls[1][0]).toBe('keyup');
            expect(addEventListener.mock.calls[2][0]).toBe('focusout');

            const onKeyDown = addEventListener.mock.calls[0][1];
            const onKeyUp = addEventListener.mock.calls[1][1];

            expect(onChange).not.toHaveBeenCalled();
            expect(onAfterChange).not.toHaveBeenCalled();

            // simulate keydown
            onKeyDown({ key: 'ArrowRight', preventDefault: () => {} });

            expect(onChange).toHaveBeenCalledTimes(1);
            expect(onChange).toHaveBeenCalledWith(1, 0);
            expect(onAfterChange).not.toHaveBeenCalled();

            // simulate keydown
            onKeyDown({ key: 'ArrowRight', preventDefault: () => {} });

            expect(onChange).toHaveBeenCalledTimes(2);
            expect(onChange).toHaveBeenCalledWith(2, 0);
            expect(onAfterChange).not.toHaveBeenCalled();

            // simulate keyup
            onKeyUp();

            expect(onChange).toHaveBeenCalledTimes(2);
            expect(onAfterChange).toHaveBeenCalledTimes(1);
            expect(onAfterChange).toHaveBeenCalledWith(2, 0);
            expect(onAfterChange.mock.invocationCallOrder[0]).toBeGreaterThan(
                onChange.mock.invocationCallOrder[1]
            );
        });

        it('should handle left and right arrow keydown events when the slider is horizontal', async () => {
            const testRenderer = renderer.create(
                <ReactSlider min={0} max={10} step={1} thumbClassName="test-thumb" />
            );
            const testInstance = testRenderer.root;
            const thumb = testInstance.findByProps({ className: 'test-thumb test-thumb-0 ' });
            const { addEventListener } = document;

            // simulate focus on thumb
            thumb.props.onFocus();
            expect(addEventListener.mock.calls[0][0]).toBe('keydown');
            const onKeyDown = addEventListener.mock.calls[0][1];

            onKeyDown({ key: 'ArrowLeft', preventDefault: () => {} });
            const valueTestOne = testRenderer.toJSON().children[2].props['aria-valuenow'];
            expect(valueTestOne).toBe(0);

            thumb.props.onFocus();
            onKeyDown({ key: 'ArrowRight', preventDefault: () => {} });
            onKeyDown({ key: 'ArrowRight', preventDefault: () => {} });

            const valueTestTwo = testRenderer.toJSON().children[2].props['aria-valuenow'];
            expect(valueTestTwo).toBe(2);
        });

        it('should handle left and right arrow keydown events when the slider is horizontal and inverted', async () => {
            const testRenderer = renderer.create(
                <ReactSlider invert min={0} max={10} step={1} thumbClassName="test-thumb" />
            );

            const testInstance = testRenderer.root;
            const thumb = testInstance.findByProps({ className: 'test-thumb test-thumb-0 ' });
            const { addEventListener } = document;

            // simulate focus on thumb
            thumb.props.onFocus();

            expect(addEventListener.mock.calls[0][0]).toBe('keydown');

            const onKeyDown = addEventListener.mock.calls[0][1];

            onKeyDown({ key: 'ArrowLeft', preventDefault: () => {} });
            onKeyDown({ key: 'ArrowLeft', preventDefault: () => {} });
            onKeyDown({ key: 'ArrowLeft', preventDefault: () => {} });

            const valueTestOne = testRenderer.toJSON().children[2].props['aria-valuenow'];
            expect(valueTestOne).toBe(3);

            onKeyDown({ key: 'ArrowRight', preventDefault: () => {} });
            onKeyDown({ key: 'ArrowRight', preventDefault: () => {} });

            const valueTestTwo = testRenderer.toJSON().children[2].props['aria-valuenow'];
            expect(valueTestTwo).toBe(1);
        });
    });

    it('should replace state value when props value changes', () => {
        const mockRenderThumb = jest.fn();
        const mockFirstValue = 40;
        const mockSecondValue = 80;
        const testRenderer = renderer.create(
            <ReactSlider
                thumbClassName="test-thumb"
                renderThumb={mockRenderThumb}
                value={mockFirstValue}
                min={0}
                step={1}
            />
        );
        expect(mockRenderThumb).toHaveBeenCalledTimes(1);
        expect(mockRenderThumb.mock.calls[0][1].value).toBe(mockFirstValue);

        renderer.act(() => {
            testRenderer.update(
                <ReactSlider
                    thumbClassName="test-thumb"
                    renderThumb={mockRenderThumb}
                    value={mockSecondValue}
                    min={0}
                    step={1}
                />
            );
        });

        expect(mockRenderThumb).toHaveBeenCalledTimes(2);
        expect(mockRenderThumb.mock.calls[1][1].value).toBe(mockSecondValue);
    });

    describe('marks prop', () => {
        describe('boolean value', () => {
            it('should not render marks if "false" is passed', () => {
                const mockRenderMark = jest.fn();
                renderer.create(
                    <ReactSlider
                        value={0}
                        min={0}
                        max={100}
                        marks={false}
                        renderMark={mockRenderMark}
                    />
                );

                expect(mockRenderMark).not.toHaveBeenCalled();
            });

            it('should render marks if "true" is passed', () => {
                const mockedMin = 0;
                const mockedMax = 10;
                const mockRenderMark = jest.fn();
                renderer.create(
                    <ReactSlider
                        value={0}
                        min={mockedMin}
                        max={mockedMax}
                        marks
                        renderMark={mockRenderMark}
                    />
                );

                expect(mockRenderMark).toHaveBeenCalledTimes(mockedMax + 1);
                mockRenderMark.mock.calls.forEach(([call], index) => {
                    expect(call.key).toBe(mockedMin + index);
                });
            });

            it('should render marks if "true" is passed and "min" is not 0', () => {
                const mockedMin = 10;
                const mockedMax = 20;
                const mockRenderMark = jest.fn();
                renderer.create(
                    <ReactSlider
                        value={0}
                        min={mockedMin}
                        max={mockedMax}
                        marks
                        renderMark={mockRenderMark}
                    />
                );

                expect(mockRenderMark).toHaveBeenCalledTimes(mockedMax - mockedMin + 1);
                mockRenderMark.mock.calls.forEach(([call], index) => {
                    expect(call.key).toBe(mockedMin + index);
                });
            });
        });

        describe('number value', () => {
            it('should render correct marks', () => {
                const mockedMin = 0;
                const mockedMax = 10;
                const mockedMarks = 2;
                const mockRenderMark = jest.fn();
                renderer.create(
                    <ReactSlider
                        value={0}
                        min={mockedMin}
                        max={mockedMax}
                        marks={mockedMarks}
                        renderMark={mockRenderMark}
                    />
                );

                expect(mockRenderMark).toHaveBeenCalledTimes(6);
                mockRenderMark.mock.calls.forEach(([call], index) => {
                    expect(call.key).toBe(mockedMin + mockedMarks * index);
                });
            });

            it('should render correct marks if "min" is not 0', () => {
                const mockedMin = 10;
                const mockedMax = 20;
                const mockedMarks = 2;
                const mockRenderMark = jest.fn();
                renderer.create(
                    <ReactSlider
                        value={0}
                        min={mockedMin}
                        max={mockedMax}
                        marks={mockedMarks}
                        renderMark={mockRenderMark}
                    />
                );

                expect(mockRenderMark).toHaveBeenCalledTimes(6);
                mockRenderMark.mock.calls.forEach(([call], index) => {
                    expect(call.key).toBe(mockedMin + mockedMarks * index);
                });
            });
        });

        describe('array of numbers is passed', () => {
            it('should render correct marks', () => {
                const mockedMin = 0;
                const mockedMax = 10;
                const mockedMarks = [1, 2, 3, 4, 5];
                const mockRenderMark = jest.fn();
                renderer.create(
                    <ReactSlider
                        value={0}
                        min={mockedMin}
                        max={mockedMax}
                        marks={mockedMarks}
                        renderMark={mockRenderMark}
                    />
                );

                expect(mockRenderMark).toHaveBeenCalledTimes(mockedMarks.length);
                mockRenderMark.mock.calls.forEach(([call], index) => {
                    expect(call.key).toBe(mockedMarks[index]);
                });
            });
        });
    });
});
