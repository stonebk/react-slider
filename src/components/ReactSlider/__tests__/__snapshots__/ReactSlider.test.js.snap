// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ReactSlider> can render 1`] = `
<div
  className="slider"
  onClick={[Function]}
  onMouseDown={[Function]}
  style={
    Object {
      "position": "relative",
    }
  }
>
  <div
    className="track track-0"
    onTouchStart={[Function]}
    style={
      Object {
        "left": 0,
        "position": "absolute",
        "right": 0,
        "touchAction": "none",
        "willChange": undefined,
      }
    }
  />
  <div
    className="track track-1"
    onTouchStart={[Function]}
    style={
      Object {
        "left": 0,
        "position": "absolute",
        "right": 0,
        "touchAction": "none",
        "willChange": undefined,
      }
    }
  />
  <div
    aria-disabled={false}
    aria-orientation="horizontal"
    aria-valuemax={100}
    aria-valuemin={0}
    aria-valuenow={0}
    className="thumb thumb-0 "
    onFocus={[Function]}
    onMouseDown={[Function]}
    onTouchStart={[Function]}
    role="slider"
    style={
      Object {
        "left": "0px",
        "position": "absolute",
        "touchAction": "none",
        "willChange": undefined,
        "zIndex": 1,
      }
    }
    tabIndex={0}
  />
</div>
`;
